### PR TITLE
Wire up trace

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following wires up the tracing at the API server level and at the database transaction runner. We still need to determine at which boundaries we want to start a span, either at a public boundary (apiserver, service, state, database transaction) or just at key critical layers (apiserver, database transaction).

There is an important change in this PR. We record the model short (`[:6]`) as a service name, instead a worker as a service name. That way we can use workers as microservices that can generate a call service graph. We may want to prefix each service name with the binary name as well to help quantify each request; `juju: apiclient -> jujud: apiserver -> jujud: dbaccessor`. I'm not wedded to this though, but to benefit from the service graph we should name things nicely.

Lastly, to prevent the need to have a new trace for every function, you can ask for the span from the context. This will then allow the addition of events and errors making it much lighter weight.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

### Setting up tempo

It assumes that docker (docker-compose) is correctly installed:

```sh
$ git clone https://github.com/grafana/tempo.git
$ cd tempo/example/docker-compose/local
$ docker compose up -d
```

### Juju

Replace `<IP ADDRESS>` with your host machine (not localhost) address (`lxc info | yq ".environment | .addresses"` is a good place to start looking)

```sh
$ juju bootstrap lxd test --build-agent --config="open-telemetry-enabled=true" --config="open-telemetry-insecure=true" --config="open-telemetry-endpoint=<IP ADDRESS>:4317"
```

### Tempo Explore

Open the following [Tempo Explore Dashboard](http://localhost:3000/explore?orgId=1&left=%7B%22datasource%22:%22tempo%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22tempo%22,%22uid%22:%22tempo%22%7D,%22queryType%22:%22nativeSearch%22,%22serviceName%22:%22apiserver%22%7D%5D,%22range%22:%7B%22from%22:%22now-1h%22,%22to%22:%22now%22%7D%7D)

Click around and have a play.

### Multiple spans

To [view multiple spans](http://localhost:3000/explore?orgId=1&left=%7B%22datasource%22:%22tempo%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22tempo%22,%22uid%22:%22tempo%22%7D,%22queryType%22:%22nativeSearch%22,%22serviceName%22:%22apiserver%22,%22spanName%22:%22txn.%28%2ARetryingTxnRunner%29.run%22%7D%5D,%22range%22:%7B%22from%22:%22now-5m%22,%22to%22:%22now%22%7D%7D), you can call `juju controller`, which will go through the dqlite database transaction runner.

Screenshot from 2023-09-20 20-40-57

## Links

**Jira card:** JUJU-[XXXX]

### CODE REVIEW

**Comment by PietroPasotti** [comment]
where's the ticket for giving the controller charm an integration with https://github.com/canonical/tempo-k8s-operator

**Comment by SimonRichardson** [comment]
@PietroPasotti we're solving logs first, then integration with tracing will come.

**Comment by SimonRichardson** [comment]
See go-dqlite library will accept the following PR: https://github.com/canonical/go-dqlite/pull/276

**Review by hmlanigan — APPROVED** [review]
By using the short of the model uuid - all of the traces look the same to me. No idea what the difference between them is beyond time. There is a guide on how to get more?

Mostly I have questions. 

QA worked as expected.

File: apiserver/export_test.go [hunk 1]
```diff
@@ -54,6 +56,10 @@ func (testingAPIRootHandler) ServiceFactory() servicefactory.ServiceFactory {
 	return nil
 }
 
+func (testingAPIRootHandler) Tracer() coretrace.Tracer {
+	return nil
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply apiserver/export_test.go:60]: It's purely for testing. We don't actually test this, just that the Tracer exists.

File: core/trace/attribute.go [hunk 2]
```diff
@@ -29,3 +31,23 @@ func (a StringAttribute) Key() string {
 func (a StringAttribute) Value() string {
 	return a.value
 }
+
+// IntAttribute defines an attribute with a string value.
+type IntAttribute struct {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply core/trace/attribute.go:36]: This is used in the next PR.

File: worker/trace/tracer.go [hunk 3]
```diff
@@ -89,18 +136,34 @@ func (t *tracer) Start(ctx context.Context, name string, opts ...coretrace.Optio
...
 	}
+	return coretrace.WithSpan(ctx, &limitedSpan{
+		Span:   managed,
+		logger: t.logger,
+	}), managed
+}
+
+// Enabled returns if the tracer is enabled.
+func (t *tracer) Enabled() bool {
+	return true
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply worker/trace/tracer.go:166]: In the next PR where we send the trace into the dqlite library, we need to know if we should bother to send the trace or not. There isn't a point in sending a potentially expensive trace down through the layers if we're not doing anything.

File: core/trace/context.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+	tracer, ok := value.(Tracer)
+	if !ok {
+		return NoopTracer{}, false
+	}
+	return tracer, tracer.Enabled()
+}
+
+// SpanFromContext returns a span from the context. If no span is found,
+// an empty span is returned.
+func SpanFromContext(ctx context.Context) Span {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply core/trace/context.go:33]: Tests, but it provides the ability to get a span without starting a trace for that layer. It's really useful as traces are expensive.

File: worker/trace/manifold.go [hunk 5]
```diff
@@ -85,13 +85,17 @@ func Manifold(config ManifoldConfig) dependency.Manifold {
 			// a noop worker. If the open telemetry does change, then we will
 			// bounce the world and this will be re-evaluated.
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply worker/trace/manifold.go:86]: This happens in the agent config updater. If you change the enable flag, we'll bounce the world.

[the remark this answers is not in the document]
Comment (SimonRichardson) [reply worker/trace/manifold.go:86]: I've added a note.

**Pull request merged**